src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <QtCore/QTimer>
+#include <QtWidgets/QMainWindow>
+#include <boost/bind.hpp>
+#include <memory>
+
+#include "software/gui/generic_widgets/robot_status/robot_status.h"
+#include "software/gui/robot_diagnostics/ui/ui_main_widget.h"
+#include "software/gui/robot_diagnostics/widgets/chicker.h"
+#include "software/gui/robot_diagnostics/widgets/dribbler.h"
+#include "software/gui/robot_diagnostics/widgets/drive.h"
+#include "software/gui/robot_diagnostics/widgets/feedback.h"
+#include "software/gui/robot_diagnostics/widgets/leds.h"
+#include "software/gui/robot_diagnostics/widgets/robot_selection.h"
+#include "software/multithreading/thread_safe_buffer.h"
+#include "software/primitive/primitive.h"
+#include "software/proto/sensor_msg.pb.h"
+
+// Forward declare the name of the top-level GUI class defined in main_widget.ui
+namespace Ui
+{
+    class AutoGeneratedMainWidget;
+}
+
+/**
+ * This is the main window for the robot diagnostics.
+ *
+ * This class uses ThreadSafeBuffers to send/receive new data, and updates the GUI with
+ * the new data at a defined rate using a timer.
+ */
+class RobotDiagnostics : public QMainWindow
+{
+    Q_OBJECT
+
+   public:
+    /**
+     * Creates a new Visualizer MainWindow
+     *
+     * @param sensor_msg_buffer The buffer used to receive new SensorMsgs
+     * @param primitive_buffer The buffer used to send new Primitives
+     * @param parent parent widget
+     */
+    explicit RobotDiagnostics(
+        std::shared_ptr<ThreadSafeBuffer<SensorMsg>> sensor_msg_buffer,
+        std::shared_ptr<ThreadSafeBuffer<std::unique_ptr<Primitive>>> primitive_buffer,
+        QWidget* parent = nullptr);
+
+   private:
+    /**
+     * Push new primitive to buffer using new values received
+     *
+     * @param new values
+     */
+    void receiveChickerState(double chicker_power, ChargeMode charge_mode,
```

Comment:
nit: These really feel more like callback functions to me. `receive` in the name makes me think we're getting data from an external source (sort of like the ThreadedObservers)

Reply:
changed to on__changed